### ISSUE TITLE
Test running strategies

### ISSUE THREAD

**Issue by theofidry** [infection/infection#88]
In order to be as efficient as possible, I would suggest to have different running strategies in order to meet different needs:

- When code coverage is provided, make use of it to know which tests to execute and in what order (already done to a certain extend I believe?)
- Special case with PhpSpec: unless doing something weird, PhpSpec is intended to have a one-one relationship between a spec file and its source file and match a certain naming convention. As such which test to execute is trivial to decide.
- No coverage: we could either bail out or execute all the tests. IMO it is perfectly viable to not have a coverage of any kind so whilst horribly slow, running all the tests for it would be an option.
- Apply multiple mutants (1)

(1) I'm expending this point separately as IMO this is a bit more complex and tricky. Right now we apply mutants one by one and for each we run them in a separate process (parallelised or not). There is a big overhead here as we need to create as many processes as mutants, and creating a process is expansive. An alternative way of doing it would be to apply several unrelated mutants (and we can deduce that from the coverage data) in the same process.

Another variant of that mode, more aggressive, would be to push it to apply multiple mutants even if they might overlap each other (e.g. multiple mutants in the same method). If the tests pass (the mutant escaped), it's a win in perf. If the mutants died, we would need to re-iterate the process with only one mutant as you can't possibly know which mutant caused the failure and that might result in a time loss. I'm not sure which balance, maybe adding some kind of conditions on what mutants can be applied together would provide a reliable way to make it faster in most cases, but that's an idea.

**Comment by BackEndTea** [comment]
> When code coverage is provided, make use of it to know which tests to execute and in what order (already done to a certain extend I believe?)

I think this is related to an issue i'm having, so i'll post it here.

I'm not too sure how the coverage is retrieved, but when using the ```@coversNothing``` annotation for integration tests, they are considered to really not cover anything for the purpose of infection.

The ```@covers``` and ```@coversNothing``` annotations are generally used to better see what things are explicitly tested, but i feel that for the purpose of this testing framework, they should be ignored. Since it doesn't really matter if the mutator got killed by an integration or unit test.

I'm not sure if there is a proper way to get the actual coverage, instead of it being filtered with ```@covers``` and ```@coversNothing```.

**Comment by paslandau** [comment]
I'd like to expand on @theofidry's "Apply multiple mutants (1)" suggestion and provide some real live examples:

# tl;dr
1. grouping multiple unrelated tests together in the same worker/process would be great due to singnificantly reduced setup costs
2. please provide a unique identifier to each process (e.g. via environment variable) so we can set up "scoped" dependencies (e.g. databases)

# 1. Test setup cost
In our application, we do a lot of Integration / End2End testing, which usually involves a certain setup cost, e.g. run migrations on a database. It's usually sufficient to run this migration once for the whole testsuite - each test still needs to take care of seeding / clearing the database to ensure it's own state, but the initial migration only needs to happen once.

We're currently keeping track of that with a static property in our base test class that gets set after the initial migration so that every subsequent test can skip that (costly) step.

> Right now we apply mutants one by one and for each we run them in a separate process (parallelised or not). 

This means in effect, that this "setup cost" is currently due on every single process. Thus, grouping would be very much appreciated. I wouldn't even keep it only to "several unrelated mutants"

> An alternative way of doing it would be to apply several unrelated mutants (and we can deduce that from the coverage data) in the same process.

but actually to several (unrelated) tests in general that are run by the same process. That's how https://github.com/paratestphp/paratest is doing it.

# 2. Unique identifiers for each process
Another side effect of parallel tests that hit the database is the requirement, that we need the ability to create "scoped" databases on the fly. Concrete example:
- Our main application uses the database with the name database name `production`
- Our tests use the database with the name database name `testing`
- When we run tests in parallel, each process uses a database named `testing_<process_identifier>`

That strategy ensures that each process will not interfere with any other parallel running process, because each process has it's own space (or "scope"). This is also useful for things like file systems or even external APIs (think AWS oder Google Cloud).

In https://github.com/paratestphp/paratest this is solved by passing an (unique) environment variable called `TEST_TOKEN` to each process (actually just an incremented number). In our application we evaluate that variable and set up the database, file system, etc. accordingly.

I didn't see anything like that for infection - hence this would also be greatly appreciated :)

**Comment by borNfreee** [comment]
Hi @paslandau, thanks for your input.

Regarding "grouping" - there are a bunch of existing [Mutant Reduction Techniques ](http://crestweb.cs.ucl.ac.uk/resources/mutation_testing_repository/technique.php) with a mathematical background that should be taken into account. Personally, I'm interested in Higher Order Mutants and the paper ["Constructing Subtle Faults Using Higher Order Mutation Testing"](https://www.researchgate.net/publication/224332299_Constructing_Subtle_Faults_Using_Higher_Order_Mutation_Testing). 

Didn't dig yet, but this is an interesting topic and will be not easy to implement I guess.

Regarding "Unique identifiers for each process": this seems like a trivial task - to exporta a unique environment variable for each Mutant. We can extract this feature request and implement it for the upcoming release (as always, PR would be much appreciated).
